drivers/bus/acpi/compbatt/compbatt.c

```diff
@@ -65,32 +67,351 @@ CompBattSystemControl(IN PDEVICE_OBJECT DeviceObject,
+                             "          Voltage -> %u\n"
+                             "          Rate -> %u\n",
+                             BatteryStatus.PowerState,
+                             BatteryStatus.Capacity,
+                             BatteryStatus.Voltage,
+                             BatteryStatus.Rate);
+                }
+
+                /*
+                 * Now determine whether should we notify the Battery Class driver due to
+                 * changes in power state settings in the composite battery. This could
+                 * happen in two following conditions:
+                 *
+                 * 1. The status notify flag was set for the respective power notification
+                 *    settings, and the composite battery incurred in a change of such
+                 *    settings. In this case we have to probe the current settings that
+                 *    they have changed.
+                 *
+                 * 2. The status notify flag was not set, therefore we do not know the
+                 *    exact configuration of the notification settings. We only care that
+                 *    the power state has changed at this point.
+                 *
+                 * Why do we have to do this is because we have to warn the Battery Class
+                 * about the data that has changed.
+                 */
+                if (!(DeviceExtension->Flags & COMPBATT_STATUS_NOTIFY_SET))
+                {
+                    if (PrevPowerState != DeviceExtension->BatteryStatus.PowerState)
+                    {
+                        /* The previous power state is no longer valid, notify Battery Class */
+                        BatteryClassStatusNotify(DeviceExtension->ClassData);
+                    }
+                }
+                else
+                {
+                    /*
+                     * Unlike the condition above, we check for power state change against
+                     * the current notify wait set since the notify set flag bit is assigned.
+                     */
+                    if (DeviceExtension->WaitNotifyStatus.PowerState != DeviceExtension->BatteryStatus.PowerState ||
+                        DeviceExtension->WaitNotifyStatus.LowCapacity > DeviceExtension->BatteryStatus.Capacity ||
+                        DeviceExtension->WaitNotifyStatus.HighCapacity < DeviceExtension->BatteryStatus.Capacity)
+                    {
+                        /* The following configuration settings have changed, notify Battery Class */
+                        BatteryClassStatusNotify(DeviceExtension->ClassData);
+                    }
+                }
+            }
+
+            break;
+        }
+
+        default:
+        {
+            if (CompBattDebug & COMPBATT_DEBUG_ERR)
+            {
+                DbgPrint("CompBatt: BAD!!! WE SHOULD NOT BE HERE!\n");
+            }
+
+            ASSERT(FALSE);
```

Comment:
Looks like it could be transformed into an `ASSERTMSG("CompBatt: BAD!!! WE SHOULD NOT BE HERE!\n", FALSE);` ? :thinking: 
Maybe also add an `UNREACHABLE;`